estimatefee.go

```diff
@@ -0,0 +1,434 @@
+	mined    int32 // The block in which this tx was mined.
+}
+
+// registeredBlock
+type registeredBlock struct {
+	hash         *chainhash.Hash
+	transactions []*observedTransaction
+}
+
+// feeEstimator is a class that manages the data necessary to create
+// fee estimations. It is safe for concurrent access.
+type feeEstimator struct {
+	maxRollback uint32
+	binSize     int
+
+	// The maximum number of replacements that can be made in a single
+	// bin per block. Default is estimateFeeMaxReplacements
+	maxReplacements int
+
+	// The minimum number of blocks that can be registered with the fee
+	// estimator before it will provide answers.
+	minRegisteredBlocks uint32
+
+	// The last known height.
+	lastKnownHeight int32
+
+	sync.RWMutex
+	observed            map[chainhash.Hash]observedTransaction
+	bin                 [estimateFeeBins][]*observedTransaction
+	numBlocksRegistered uint32 // The number of blocks that have been registered.
+
+	// The cached estimates.
+	cached []float64
+
+	// Transactions that have been removed from the bins. This allows us to
+	// revert in case of an orphaned block.
+	dropped []registeredBlock
+}
+
+// newFeeEstimator creates a feeEstimator for which at most maxRollback blocks
+// can be unregistered and which returns an error unless minRegisteredBlocks
+// have been registered with it.
+func newFeeEstimator(maxRollback, minRegisteredBlocks uint32) *feeEstimator {
+	return &feeEstimator{
+		maxRollback:         maxRollback,
+		minRegisteredBlocks: minRegisteredBlocks,
+		lastKnownHeight:     mempoolHeight,
+		binSize:             estimateFeeBinSize,
+		maxReplacements:     estimateFeeMaxReplacements,
+		observed:            make(map[chainhash.Hash]observedTransaction),
+		dropped:             make([]registeredBlock, 0, maxRollback),
+	}
+}
+
+// ObserveTransaction is called when a new transaction is observed in the mempool.
+func (ef *feeEstimator) ObserveTransaction(t *mempoolTxDesc) {
+	ef.Lock()
+	defer ef.Unlock()
+
+	hash := *t.Tx.Hash()
+	if _, ok := ef.observed[hash]; !ok {
+		size := uint32(t.Tx.MsgTx().SerializeSize())
+		fee := uint64(t.Fee)
+
+		ef.observed[hash] = observedTransaction{
+			hash:     hash,
+			size:     size,
+			fee:      fee,
+			feePerKb: float64(1000*fee) / float64(size),
+			observed: t.Height,
+			mined:    mempoolHeight,
+		}
+	}
+}
+
+// RegisterBlock informs the fee estimator of a new block to take into account.
+func (ef *feeEstimator) RecordBlock(block *btcutil.Block) {
+	ef.Lock()
+	defer ef.Unlock()
+
+	// The previous sorted list is invalid, so delete it.
+	ef.cached = nil
+
+	height := block.Height()
+	if height != ef.lastKnownHeight+1 && ef.lastKnownHeight != mempoolHeight {
+		panic(fmt.Sprint("intermediate block not recorded; current height is ", ef.lastKnownHeight,
+			"; new height is ", height))
+	}
+
+	// Update the last known height.
+	ef.lastKnownHeight = height
+	ef.numBlocksRegistered++
+
+	// Randomly order txs in block.
```

Comment:
It's because only we only make a maximum of estimateFeeMaxReplacements (default 10) per bin per registered block. If the transactions were not randomly ordered, we would bias the transactions toward those which are in the block first. There is no reason to assume that the first transactions in a block can be treated like a random sample.